frontend/src/queries/preferencesQueries.ts

```diff
@@ -0,0 +1,492 @@
+      if (context?.previousData) {
+        queryClient.setQueryData(
+          preferencesQueryKeys.all,
+          context.previousData
+        );
+      }
+    },
+    // Mark as stale but don't force refetch - respect staleTime
+    onSettled: () => {
+      queryClient.invalidateQueries({
+        queryKey: preferencesQueryKeys.all,
+        refetchType: 'none'
+      });
+    }
+  });
+}
+
+/**
+ * Mutation hook for updating favorites (zone, fileSharePath, or folder)
+ *
+ * @example
+ * const mutation = useUpdateFavoritesMutation();
+ * mutation.mutate({ preferenceKey: 'zone', updatedMap: { ... } });
+ */
+export function useUpdateFavoritesMutation(): UseMutationResult<
+  void,
+  Error,
+  FavoriteUpdatePayload,
+  { previousData?: PreferencesApiResponse }
+> {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: async (
+      { preferenceKey, updatedMap }: FavoriteUpdatePayload,
+      { signal }
+    ) => {
+      const response = await sendFetchRequest(
+        `/api/preference/${preferenceKey}`,
+        'PUT',
+        { value: Object.values(updatedMap) },
+        { signal }
+      );
+      if (!response.ok) {
+        throw await toHttpError(response);
+      }
+      // Don't return the Response object - it can't be cloned by devtools
+      return;
+    },
+    // Optimistic update for immediate UI feedback
+    onMutate: async ({ preferenceKey, updatedMap }) => {
+      // Cancel any outgoing refetches
+      await queryClient.cancelQueries({ queryKey: preferencesQueryKeys.all });
+
+      // Get the raw API response data (before transformation)
+      const previousData = queryClient.getQueryData<PreferencesApiResponse>(
+        preferencesQueryKeys.all
+      );
+
+      // Optimistically update the raw API response cache
+      queryClient.setQueryData<PreferencesApiResponse>(
+        preferencesQueryKeys.all,
+        old => {
+          if (!old) {
+            return old;
+          }
+          // Update the appropriate favorites array in raw format
+          return {
+            ...old,
+            [preferenceKey]: { value: Object.values(updatedMap) }
+          };
+        }
+      );
+
+      return { previousData };
+    },
+    // On error, rollback
+    onError: (_err, _variables, context) => {
+      if (context?.previousData) {
+        queryClient.setQueryData(
+          preferencesQueryKeys.all,
+          context.previousData
+        );
+      }
+    },
+    // Mark as stale but don't force refetch - respect staleTime
+    onSettled: () => {
+      queryClient.invalidateQueries({
+        queryKey: preferencesQueryKeys.all,
+        refetchType: 'none'
+      });
+    }
+  });
+}
+
+/**
+ * Mutation hook for updating recently viewed folders with optimistic updates
+ *
+ * @example
+ * const mutation = useUpdateRecentlyViewedFoldersMutation();
+ * mutation.mutate(updatedFolders);
+ */
+export function useUpdateRecentlyViewedFoldersMutation(): UseMutationResult<
```

Comment:
It looks like this and `useUpdateFavoritesMutation` are basically the same code duplicated. Can they be combined?